From a057c0bb3abae73aa636a228d766e30350b3dd48 Mon Sep 17 00:00:00 2001
From: Barracuda Team <admin@omega8.cc>
Date: Mon, 9 Dec 2013 14:14:17 +0100
Subject: [PATCH] Issue #1559116 by mva.name, omega8cc - Make core
 aware of Nginx and PHP-FPM to avoid confusing alerts.

---
 includes/file.inc             | 20 ++++++++++++++++++++
 modules/file/file.install     | 28 ++++++++++++++++++++++------
 modules/system/system.install | 20 ++++++++++++++++++--
 3 files changed, 60 insertions(+), 8 deletions(-)

diff --git a/includes/file.inc b/includes/file.inc
index 0ec69b7..ab27b82 100644
--- a/includes/file.inc
+++ b/includes/file.inc
@@ -475,6 +475,26 @@ function file_ensure_htaccess() {
  *   already present. Defaults to FALSE.
  */
 function file_create_htaccess($directory, $private = TRUE, $force_overwrite = FALSE) {
+
+  // Test the web server identity.
+  if (!isset($_SERVER['SERVER_SOFTWARE'])) {
+    // Skip this in Drush where SERVER_SOFTWARE is not set and .htaccess never used anyway.
+    $use_htaccess = FALSE;
+  }
+  elseif (isset($_SERVER['SERVER_SOFTWARE']) && preg_match("/Nginx/i", $_SERVER['SERVER_SOFTWARE'])) {
+    $use_htaccess = FALSE;
+  }
+  elseif (isset($_SERVER['SERVER_SOFTWARE']) && preg_match("/Apache/i", $_SERVER['SERVER_SOFTWARE'])) {
+    $use_htaccess = TRUE;
+  }
+  else {
+    $use_htaccess = FALSE;
+  }
+
+  if (!$use_htaccess) {
+    return;
+  }
+
   if (file_uri_scheme($directory)) {
     $directory = file_stream_wrapper_uri_normalize($directory);
   }
diff --git a/modules/file/file.install b/modules/file/file.install
index 47ee4fd..d3b796f 100644
--- a/modules/file/file.install
+++ b/modules/file/file.install
@@ -53,18 +53,34 @@ function file_requirements($phase) {

   // Check the server's ability to indicate upload progress.
   if ($phase == 'runtime') {
-    $implementation = file_progress_implementation();
-    $apache = strpos($_SERVER['SERVER_SOFTWARE'], 'Apache') !== FALSE;
-    $fastcgi = strpos($_SERVER['SERVER_SOFTWARE'], 'mod_fastcgi') !== FALSE || strpos($_SERVER["SERVER_SOFTWARE"], 'mod_fcgi') !== FALSE;
     $description = NULL;
-    if (!$apache) {
+    $implementation = file_progress_implementation();
+
+    // Test the web server identity.
+    if (isset($_SERVER['SERVER_SOFTWARE']) && preg_match("/Nginx/i", $_SERVER['SERVER_SOFTWARE'])) {
+      $is_nginx = TRUE;
+      $is_apache = FALSE;
+      $fastcgi = FALSE;
+    }
+    elseif (isset($_SERVER['SERVER_SOFTWARE']) && preg_match("/Apache/i", $_SERVER['SERVER_SOFTWARE'])) {
+      $is_nginx = FALSE;
+      $is_apache = TRUE;
+      $fastcgi = strpos($_SERVER['SERVER_SOFTWARE'], 'mod_fastcgi') !== FALSE || strpos($_SERVER["SERVER_SOFTWARE"], 'mod_fcgi') !== FALSE;
+    }
+    else {
+      $is_nginx = FALSE;
+      $is_apache = FALSE;
+      $fastcgi = FALSE;
+    }
+
+    if (!$is_apache && !$is_nginx) {
       $value = t('Not enabled');
-      $description = t('Your server is not capable of displaying file upload progress. File upload progress requires an Apache server running PHP with mod_php.');
+      $description = t('Your server is not capable of displaying file upload progress. File upload progress requires an Apache server running PHP with mod_php or Nginx with PHP-FPM.');
       $severity = REQUIREMENT_INFO;
     }
     elseif ($fastcgi) {
       $value = t('Not enabled');
-      $description = t('Your server is not capable of displaying file upload progress. File upload progress requires PHP be run with mod_php and not as FastCGI.');
+      $description = t('Your server is not capable of displaying file upload progress. File upload progress requires PHP be run with mod_php or PHP-FPM and not as FastCGI.');
       $severity = REQUIREMENT_INFO;
     }
     elseif (!$implementation && extension_loaded('apc')) {
diff --git a/modules/system/system.install b/modules/system/system.install
index 7d3c959..8433faf 100644
--- a/modules/system/system.install
+++ b/modules/system/system.install
@@ -253,8 +253,24 @@ function system_requirements($phase) {
     $requirements['settings.php']['title'] = $t('Configuration file');
   }

-  // Test the contents of the .htaccess files.
-  if ($phase == 'runtime') {
+  // Test the web server identity.
+  if (!isset($_SERVER['SERVER_SOFTWARE'])) {
+    // Skip this in Drush where SERVER_SOFTWARE is not set and .htaccess never used anyway.
+    $use_htaccess = FALSE;
+  }
+  elseif (isset($_SERVER['SERVER_SOFTWARE']) && preg_match("/Nginx/i", $_SERVER['SERVER_SOFTWARE'])) {
+    $use_htaccess = FALSE;
+  }
+  elseif (isset($_SERVER['SERVER_SOFTWARE']) && preg_match("/Apache/i", $_SERVER['SERVER_SOFTWARE'])) {
+    $use_htaccess = TRUE;
+  }
+  else {
+    $use_htaccess = FALSE;
+  }
+
+  // Test the contents of the .htaccess files only for Apache web server.
+  // The .htaccess is never used in other popular web servers, like Nginx.
+  if ($phase == 'runtime' && $use_htaccess) {
     // Try to write the .htaccess files first, to prevent false alarms in case
     // (for example) the /tmp directory was wiped.
     file_ensure_htaccess();
--
1.8.3.4 (Apple Git-47)
